blazar/plugins/devices/k3s_plugin.py

```diff
@@ -0,0 +1,184 @@
+        '''Get if the node has label=value, or if value is none, any value'''
+        return (label in node.metadata.labels and
+                (value is None or node.metadata.labels.get(label) == value))
+
+    def set_label(self, name, label, value):
+        body = {
+            "metadata": {
+                "labels": {
+                    label: value,
+                },
+            },
+        }
+        return self.core_v1.patch_node(name, body)
+
+    def set_res_id_label(self, name, reservation_id):
+        return self.set_label(
+            name, LABELS["reservation_id"], reservation_id)
+
+    def set_device(self, name, value):
+        return self.set_label(name, LABELS["device"], value)
+
+    def get_node_by_name(self, name):
+        node_list = self.core_v1.list_node()
+        for node in node_list.items:
+            if node.metadata.name == name:
+                return node
+        return node
+
+    def get_nodes_by_label(self, label, value):
+        return [
+            node for node in self.core_v1.list_node().items
+            if self.has_label(node, label, value)
+        ]
+
+    def create_device(self, device_values):
+        device_name = device_values.get('name')
+
+        if device_name is None:
+            raise manager_ex.InvalidHost(host=device_values)
+
+        node = self.get_node_by_name(device_name)
+        if not node:
+            raise manager_ex.DeviceNotFound(device=device_name)
+
+        device_properties = {
+            'name': device_name,
+            'device_type': 'container',
+            'device_driver': K3sPlugin.device_driver
+        }
+
+        to_store = set(device_values.keys()) - set(device_properties.keys())
+        extra_capabilities_keys = to_store
+        extra_capabilities = dict(
+            (key, device_values[key]) for key in extra_capabilities_keys
+        )
+
+        if any([len(key) > 64 for key in extra_capabilities_keys]):
+            raise manager_ex.ExtraCapabilityTooLong()
+
+        device = None
+        cantaddextracapability = []
+        try:
+            device = db_api.device_create(device_properties)
+        except db_ex.BlazarDBException:
+            raise
+        for key in extra_capabilities:
+            values = {'device_id': device['id'],
+                      'capability_name': key,
+                      'capability_value': extra_capabilities[key],
+                      }
+            try:
+                db_api.device_extra_capability_create(values)
+            except db_ex.BlazarDBException:
+                cantaddextracapability.append(key)
+        if cantaddextracapability:
+            raise manager_ex.CantAddExtraCapability(
+                keys=cantaddextracapability,
+                host=device['id'])
+        self.set_device(device_name, device['id'])
+        return device['id']
+
+    def is_active(self, node):
+        conditions = node.status.conditions
+        for i in conditions:
```

Comment:
Is it possible this is a sequence of conditions that can be like `NotReady... Ready... NotReady`? As in, it could go from ready to not-ready but this code would still think the node is active?

Reply:
My understanding is that the condition `Ready` is either `True` or something else. When `kubectl` shows `Not Ready` it means that `Ready` is `False`. There should only be one condition of the type `Ready`, as far as I understand.

https://kubernetes.io/docs/concepts/architecture/nodes/#condition

Reply:
Gotcha, thanks!